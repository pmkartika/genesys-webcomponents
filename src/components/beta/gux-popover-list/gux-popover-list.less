@import (reference) '../../../style/color.less';
@import (reference) '../../../style/shadows.less';
@import (reference) '../../../style/spacing.less';

@gux-arrow-height: 5px;

:host {
  z-index: 2;
  color: @gux-black-50;
}

.gux-popover-wrapper {
  display: inline-block;
  padding: @spacing-xs 0;
  background-color: @gux-grey-100;
  border: 1px solid @gux-grey-30;
  border-radius: @spacing-2xs;
  .gux-shadow-20();

  .gux-arrow,
  .gux-arrow::before {
    position: absolute;
    width: @gux-arrow-height * 2;
    height: @gux-arrow-height * 2;
    background: inherit;
  }

  .gux-arrow {
    visibility: hidden;
  }

  .gux-arrow::before {
    visibility: visible;
    content: '';
    transform: rotate(45deg);
  }

  &[data-popper-placement='bottom'],
  &[data-popper-placement='bottom-start'],
  &[data-popper-placement='bottom-end'] {
    .gux-arrow {
      top: -1 * (@gux-arrow-height);
    }

    .gux-arrow::before {
      border-top: 1px solid @gux-grey-30;
      border-left: 1px solid @gux-grey-30;
    }
  }

  &[data-popper-placement='top'],
  &[data-popper-placement='top-start'],
  &[data-popper-placement='top-end'] {
    .gux-arrow {
      bottom: -1 * (@gux-arrow-height);
    }

    .gux-arrow::before {
      border-right: 1px solid @gux-grey-30;
      border-bottom: 1px solid @gux-grey-30;
    }
  }

  &[data-popper-placement='left'],
  &[data-popper-placement='left-start'],
  &[data-popper-placement='left-end'] {
    .gux-arrow {
      right: -1 * (@gux-arrow-height);
    }

    .gux-arrow::before {
      border-top: 1px solid @gux-grey-30;
      border-right: 1px solid @gux-grey-30;
    }
  }

  &[data-popper-placement='right'],
  &[data-popper-placement='right-start'],
  &[data-popper-placement='right-end'] {
    .gux-arrow {
      left: -1 * (@gux-arrow-height);
    }

    .gux-arrow::before {
      border-bottom: 1px solid @gux-grey-30;
      border-left: 1px solid @gux-grey-30;
    }
  }
}
