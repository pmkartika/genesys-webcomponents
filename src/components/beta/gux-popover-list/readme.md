# gux-popover-list

<!-- Auto Generated Below -->


## Properties

| Property               | Attribute                | Description                                                                          | Type                                                                                                                                                                 | Default     |
| ---------------------- | ------------------------ | ------------------------------------------------------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `closeOnClickOutside`  | `close-on-click-outside` | Close popover when the user clicks outside of its bounds                             | `boolean`                                                                                                                                                            | `false`     |
| `displayDismissButton` | `display-dismiss-button` | Indicate if the dismiss button is displayed                                          | `boolean`                                                                                                                                                            | `undefined` |
| `for`                  | `for`                    | Indicates the id of the element the popover should anchor to                         | `string`                                                                                                                                                             | `undefined` |
| `position`             | `position`               | Indicate position of popover element arrow (follow popper js position attribute api) | `"bottom" \| "bottom-end" \| "bottom-start" \| "left" \| "left-end" \| "left-start" \| "right" \| "right-end" \| "right-start" \| "top" \| "top-end" \| "top-start"` | `'bottom'`  |


## Events

| Event        | Description                             | Type                |
| ------------ | --------------------------------------- | ------------------- |
| `guxdismiss` | Fired when a user dismisses the popover | `CustomEvent<void>` |


## Dependencies

### Used by

 - [gux-tab-advanced](../../stable/gux-tabs-advanced/gux-tab-advanced)
 - [gux-tab-legacy](../../legacy/gux-tabs-legacy/gux-tab-legacy)

### Depends on

- [gux-dismiss-button](../../stable/gux-dismiss-button)

### Graph
```mermaid
graph TD;
  gux-popover-list --> gux-dismiss-button
  gux-dismiss-button --> gux-icon
  gux-tab-advanced --> gux-popover-list
  gux-tab-legacy --> gux-popover-list
  style gux-popover-list fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
