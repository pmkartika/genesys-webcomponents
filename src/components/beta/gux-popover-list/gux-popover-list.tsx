import { createPopper, Instance } from '@popperjs/core';
import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
  State,
  Watch
} from '@stencil/core';

import { OnClickOutside } from '../../../utils/decorator/on-click-outside';
import { onHiddenChange } from '../../../utils/dom/on-attribute-change';
import { trackComponent } from '../../../usage-tracking';

import { PopperPosition } from './gux-popover-list.types';

@Component({
  styleUrl: 'gux-popover-list.less',
  tag: 'gux-popover-list',
  shadow: true
})
export class GuxPopoverList {
  private popperInstance: Instance;
  private hiddenObserver: MutationObserver;
  private popupElement: HTMLDivElement;

  @Element()
  private root: HTMLElement;

  /**
   * Indicates the id of the element the popover should anchor to
   */
  @Prop()
  for: string;

  /**
   * Indicate position of popover element arrow (follow popper js position attribute api)
   */
  @Prop()
  position: PopperPosition = 'bottom';

  /**
   * Indicate if the dismiss button is displayed
   */
  @Prop()
  displayDismissButton: boolean;

  /**
   * Close popover when the user clicks outside of its bounds
   */
  @Prop()
  closeOnClickOutside: boolean = false;

  /**
   * Fired when a user dismisses the popover
   */
  @Event()
  guxdismiss: EventEmitter<void>;

  @State()
  hidden: boolean = true;

  @Watch('hidden')
  hiddenHandler(hidden: boolean) {
    if (!hidden && !this.popperInstance) {
      this.runPopper();
    } else if (!hidden && this.popperInstance) {
      this.popperInstance.forceUpdate();
    }
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  checkForClickOutside(event: MouseEvent) {
    const clickPath = event.composedPath();
    const forElement = document.getElementById(this.for);
    const clickedForElement = clickPath.includes(forElement);

    if (
      (this.closeOnClickOutside || !this.displayDismissButton) &&
      !this.hidden &&
      !clickedForElement
    ) {
      this.dismiss();
    }
  }

  private runPopper(): void {
    const forElement = document.getElementById(this.for);

    if (!forElement) {
      console.error(
        `gux-popover-list: invalid "for" attribute. No element in page with the id "${this.for}"`
      );
    } else if (this.popupElement) {
      this.popperInstance = createPopper(forElement, this.popupElement, {
        modifiers: [
          {
            name: 'offset',
            options: {
              offset: [0, 7]
            }
          },
          {
            name: 'arrow',
            options: {
              padding: 16
            }
          }
        ],
        placement: this.position
      });
    }
  }

  private destroyPopper(): void {
    if (this.popperInstance) {
      this.popperInstance.destroy();
      this.popperInstance = null;
    }
  }

  private dismiss(): void {
    const dismissEvent = this.guxdismiss.emit();
    if (!dismissEvent.defaultPrevented) {
      this.root.setAttribute('hidden', '');
    }
  }

  connectedCallback(): void {
    trackComponent(this.root, { variant: this.position });

    this.hiddenObserver = onHiddenChange(this.root, (hidden: boolean) => {
      this.hidden = hidden;
    });

    this.hidden = this.root.hidden;
  }

  componentDidLoad(): void {
    this.runPopper();
  }

  disconnectedCallback(): void {
    this.destroyPopper();
    this.hiddenObserver.disconnect();
  }

  render(): JSX.Element {
    return (
      <div
        ref={(el: HTMLDivElement) => (this.popupElement = el)}
        class="gux-popover-wrapper"
      >
        <div class="gux-arrow" data-popper-arrow />
        {this.displayDismissButton && (
          <gux-dismiss-button
            onClick={this.dismiss.bind(this)}
          ></gux-dismiss-button>
        )}
        <div class="gux-popover-content">
          <slot />
        </div>
      </div>
    ) as JSX.Element;
  }
}
