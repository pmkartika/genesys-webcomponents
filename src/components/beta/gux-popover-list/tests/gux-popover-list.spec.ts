import { newSpecPage } from '@stencil/core/testing';
import * as popperjs from '@popperjs/core';
import MutationObserver from 'mutation-observer';
import { GuxPopoverList } from '../gux-popover-list';

const components = [GuxPopoverList];
const language = 'en';

describe('gux-popover-list', () => {
  beforeEach(async () => {
    global.MutationObserver = MutationObserver;
    // popperjs does not work with Stencils MockHTMLElements used in tests
    jest.spyOn(popperjs, 'createPopper').mockReturnValue({
      destroy: jest.fn()
    } as unknown as popperjs.Instance);
  });

  afterEach(async () => {
    jest.spyOn(popperjs, 'createPopper').mockRestore();
  });

  describe('#render', () => {
    [
      {
        description: 'should render popover',
        html: `
          <div>
            <div id="popover-target">
              Example Element
            </div>
            <gux-popover-list id="popover-example" position="top" for="popover-target">
              <div>popover content</div>
            </gux-popover-list>
          </div>
        `
      }
    ].forEach(({ description, html }) => {
      it(description, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.rootInstance).toBeInstanceOf(GuxPopoverList);

        expect(page.root).toMatchSnapshot();
      });
    });
  });
});
