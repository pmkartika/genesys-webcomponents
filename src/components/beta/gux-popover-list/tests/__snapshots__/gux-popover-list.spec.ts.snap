// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-popover-list #render should render popover 1`] = `
<gux-popover-list for="popover-target" id="popover-example" position="top">
  <mock:shadow-root>
    <div class="gux-popover-wrapper">
      <div class="gux-arrow" data-popper-arrow></div>
      <div class="gux-popover-content">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <div>
    popover content
  </div>
</gux-popover-list>
`;
